File: python/scripts/week-02/demo-06-prompt-engineering.py
"""
Demo 6: Prompt Engineering + Promptfoo Evaluation

What it shows:
  Same task, same schema, three prompt variants:
  - zero-shot
  - few-shot
  - chain-of-thought

The goal is to compare their behavior, especially schema adherence and field quality.

Run: python scripts/week-02/demo-06-prompt-engineering.py
"""

import os
import sys
from pathlib import Path

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))

from src.llm_functions import analyze_pr_with_prompt
from src.config import DEVBUDDY_MODEL

PROMPT_DIR = Path(__file__).resolve().parents[3] / "shared" / "prompts"

PR_TITLE = "Fix login redirect loop in auth-service"
PR_DIFF = (
    "Updated session token validation in auth.py. Expired tokens now return 401 instead of looping.\n"
    "Files: src/auth.py, tests/test_auth.py"
)


def load_prompt(name: str) -> str:
    return (PROMPT_DIR / f"{name}.txt").read_text(encoding="utf-8").strip()


def print_result(label: str, result) -> None:
    print(f"\n[{label}]")
    print(f"  project: {result.project}")
    print(f"  severity: {result.severity}")
    print(f"  summary: {result.summary}")
    print(f"  affected_files: {result.affected_files}")


def main() -> None:
    prompts = {
        "zero-shot": load_prompt("zero-shot"),
        "few-shot": load_prompt("few-shot"),
        "chain-of-thought": load_prompt("chain-of-thought"),
    }

    print("=" * 80)
    print("  Demo 6: Prompt Engineering + Promptfoo Evaluation")
    print("=" * 80)
    print(f"  Model: {DEVBUDDY_MODEL}")
    print("  Task: same PR, same schema, different prompt strategy")
    print()

    for name, template in prompts.items():
        try:
            result = analyze_pr_with_prompt(
                title=PR_TITLE,
                diff=PR_DIFF,
                prompt_template=template,
                temperature=0.0,
                max_tokens=512,
            )
            print_result(name, result)
        except Exception as exc:
            print(f"\n[{name}] ERROR: {exc}")

    print("\n" + "-" * 80)
    print("  Why this matters:")
    print("  - zero-shot is the shortest, but often least robust")
    print("  - few-shot adds examples and improves format adherence")
    print("  - chain-of-thought may improve reasoning but can be verbose")
    print("  Promptfoo helps compare these systematically across models and prompts.")
    print("-" * 80)


if __name__ == "__main__":
    main()

File: nodejs/scripts/week-02/demo-06-prompt-engineering.js
/**
 * Demo 6: Prompt Engineering + Promptfoo Evaluation
 *
 * Same PR, same Zod schema, three prompt variants:
 * - zero-shot
 * - few-shot
 * - chain-of-thought
 *
 * Run: node scripts/week-02/demo-06-prompt-engineering.js
 */
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { analyzePrWithPrompt } from "../../src/llm_functions.js";
import { config } from "../../src/config.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const promptDir = path.resolve(__dirname, "../../../shared/prompts");

const PR_TITLE = "Fix login redirect loop in auth-service";
const PR_DIFF = [
  "Updated session token validation in auth.py. Expired tokens now return 401 instead of looping.",
  "Files: src/auth.py, tests/test_auth.py",
].join("\n");

function loadPrompt(name) {
  return fs.readFileSync(path.join(promptDir, `${name}.txt`), "utf8").trim();
}

function printResult(label, result) {
  console.log(`\n[${label}]`);
  console.log(`  project: ${result.project}`);
  console.log(`  severity: ${result.severity}`);
  console.log(`  summary: ${result.summary}`);
  console.log(`  affected_files: ${JSON.stringify(result.affected_files)}`);
}

console.log("=".repeat(80));
console.log("  Demo 6: Prompt Engineering + Promptfoo Evaluation");
console.log("=".repeat(80));
console.log(`  Model: ${config.model}`);
console.log("  Task: same PR, same schema, different prompt strategy");
console.log();

for (const [name, template] of Object.entries({
  "zero-shot": loadPrompt("zero-shot"),
  "few-shot": loadPrompt("few-shot"),
  "chain-of-thought": loadPrompt("chain-of-thought"),
})) {
  try {
    const result = await analyzePrWithPrompt({
      title: PR_TITLE,
      diff: PR_DIFF,
      promptTemplate: template,
      temperature: 0.0,
      maxTokens: 512,
    });
    printResult(name, result);
  } catch (error) {
    console.log(`\n[${name}] ERROR: ${error.message}`);
  }
}

console.log("\n" + "-".repeat(80));
console.log("  Why this matters:");
console.log("  - zero-shot is the shortest, but often least robust");
console.log("  - few-shot adds examples and improves format adherence");
console.log("  - chain-of-thought may improve reasoning but can be verbose");
console.log("  Promptfoo helps compare these systematically across models and prompts.");
console.log("-".repeat(80));
